producer/writer/consumer_service_writer.go

```diff
@@ -0,0 +1,311 @@
+		consumerWriters: make(map[string]consumerWriter),
+		isClosed:        false,
+		doneCh:          make(chan struct{}),
+		m:               newConsumerServiceWriterMetrics(opts.InstrumentOptions().MetricsScope()),
+	}
+	w.processPlacementFn = w.processPlacement
+	return w, nil
+}
+
+func initShardWriters(
+	router ackRouter,
+	cs topic.ConsumerService,
+	numberOfShards uint32,
+	mPool messagePool,
+	opts Options,
+) []shardWriter {
+	sws := make([]shardWriter, numberOfShards)
+	for i := range sws {
+		switch cs.ConsumptionType() {
+		case topic.Shared:
+			sws[i] = newSharedShardWriter(uint32(i), router, mPool, opts)
+		case topic.Replicated:
+			sws[i] = newReplicatedShardWriter(uint32(i), numberOfShards, router, mPool, opts)
+		}
+	}
+	return sws
+}
+
+func (w *consumerServiceWriterImpl) Write(d producer.RefCountedData) error {
+	shard := d.Shard()
+	if shard >= w.numberOfShards {
+		w.m.invalidShard.Inc(1)
+		return fmt.Errorf("could not write data for shard %d which is larger than max shard id %d", shard, w.numberOfShards)
+	}
+	if d.Filter(w.dataFilter) {
+		w.shardWriters[shard].Write(d)
+	}
+	// It is not an error if the data does not pass the filter.
+	return nil
+}
+
+func (w *consumerServiceWriterImpl) Init() error {
+	var (
+		pw      placement.Watch
+		err     error
+		retrier = retry.NewRetrier(w.opts.PlacementWatchRetryOptions().SetForever(true))
+	)
+	retrier.Attempt(func() error {
+		pw, err = w.ps.Watch()
+		if err != nil {
+			// Unexpected, but if it ever happens, we will retry forever to recreate the watch.
+			w.logger.Errorf(
+				"could not create placement watch for %s, retry later: %v",
+				w.cs.ServiceID().String(),
+				err,
+			)
+			return err
+		}
+		return nil
+	})
+	err = w.waitForInitialPlacement(pw)
+
+	w.wg.Add(1)
+	go func() {
+		w.watchPlacementUpdatesForever(pw)
+		w.wg.Done()
+	}()
+	return err
+}
+
+func (w *consumerServiceWriterImpl) waitForInitialPlacement(pw placement.Watch) error {
+	timeout := w.opts.PlacementWatchInitTimeout()
+	select {
+	case <-pw.C():
+		return w.processPlacementWatchUpdate(pw)
+	case <-time.After(timeout):
+		return fmt.Errorf("could not receive initial placement for %s within %v", w.cs.ServiceID().String(), timeout)
+	}
+}
+
+func (w *consumerServiceWriterImpl) watchPlacementUpdatesForever(pw placement.Watch) {
+	for {
+		select {
+		case <-pw.C():
+			w.processPlacementWatchUpdate(pw)
+		case <-w.doneCh:
+			for _, cw := range w.consumerWriters {
+				cw.Close()
+			}
+			return
+		}
+	}
+}
+
+func (w *consumerServiceWriterImpl) processPlacementWatchUpdate(pw placement.Watch) error {
+	p, err := pw.Get()
+	if err != nil {
+		w.m.placementError.Inc(1)
+		w.logger.Errorf("could not process placement update from kv, %v", err)
+		return err
+	}
+	w.m.placementUpdate.Inc(1)
+	w.processPlacementFn(p)
+	return nil
+}
+
+func (w *consumerServiceWriterImpl) processPlacement(p placement.Placement) {
+	// NB(cw): Lock is not needed as w.consumerWriters is only accessed in this thread.
```

Comment:
Honestly I'd just put a lock here, it's difficult to keep track of which thread accesses what, and the frequency of placement updates is so low the locking overhead is rarely a problem.